lib/gringotts/gateways/sagepay.ex

```diff
@@ -0,0 +1,317 @@
+     aliases.
+
+  We'll be using these in the examples below.
+  [gs]: file:///home/anwar/gringotts/doc/Gringotts.Gateways.SagePay.html#
+  [example]: https://github.com/aviabird/gringotts_example
+  [iex-docs]: https://hexdocs.pm/iex/IEx.html#module-the-iex-exs-file
+  [sagepay.iex.exs]: https://github.com/Anwar0902/graph/blob/master/sagepay.iex.exs
+  """
+
+  # The Base module has the (abstract) public API, and some utility
+  # implementations.
+  use Gringotts.Gateways.Base
+
+  # The Adapter module provides the `validate_config/1`
+  # Add the keys that must be present in the Application config in the
+  # `required_config` list
+  use Gringotts.Adapter, required_config: []
+  alias Gringotts.{Money, CreditCard, Response}
+  @url "https://pi-test.sagepay.com/api/v1/"
+
+  # SagePay supports payment only by providing Merchant `:auth_id` and `:merchant_name`
+  # which generates `merchant_session_key` and then providing card details generates
+  # `card_identifier` which is required for authorization.
+
+  @doc """
+  Performs a (pre) Authorize operation.
+
+  The authorization validates the `card` details with the banking network,
+  places a hold on the transaction `amount` in the customer’s issuing bank.
+
+  SagePay returns an transaction ID(available in `Response.id`) string which can be used to:
+  * `capture/3` an amount.
+  * `void/2` abort deferred transaction.
+  * `refund/3` an amount.
+
+  ## Note
+
+  * The `merchant_session_key` expires after 400 seconds and can only be used to create one successful `card_identifier`. 
+  * It will also expire and will be removed after 3 failed attempts to create a `card_identifier`.
+  * `vendor_tx_code` in opts should always be unique.
+  * `address` should be provided as defined in `Gringotts.address`.
+
+  ## Example
+
+  The following example shows how one would (pre) authorize a payment of 42£ on
+  a sample `card`.
+
+      iex> amount = Money.new(42, :GBP)
+      iex> address = %Address{ street1: "407 St.", street2: "John Street", city: "London", postalCode: "EC1V 4AB", country: "GB"} 
+      iex> card = %Gringotts.CreditCard{number: "4929000005559",month: 3,year: 20,first_name: "SAM",last_name: "JONES",verification_code: "123",brand: "VISA"}
+      iex> opts = [
+                  config: %{
+                      auth_id: "aEpZeHN3N0hMYmo0MGNCOHVkRVM4Q0RSRkxodUo4RzU0TzZyRHBVWHZFNmhZRHJyaWE6bzJpSFNyRnliWU1acG1XT1FNdWhzWFA1MlY0ZkJ0cHVTRHNocktEU1dzQlkxT2lONmh3ZDlLYjEyejRqNVVzNXU=",
+                      merchant_name: "sandbox"
+                  },
+                  transaction_type: "Deferred",
+                  vendor_tx_code: "demotransaction-51",
+                  description: "Demo Payment",
+                  customer_first_name: "Sam",
+                  customer_last_name: "Jones",
+                  billing_address: address
+            ]
+      iex> {:ok, auth_result} = Gringotts.authorize(Gringotts.Gateways.SagePay, amount, card, opts)
```

Comment:
This is the response I received:
```elixir
iex(14)> Poison.decode! auth_result.raw
%{
  "3DSecure" => %{"status" => "CardNotEnrolled"},
  "amount" => %{"saleAmount" => 42, "surchargeAmount" => 0, "totalAmount" => 42},
  "avsCvcCheck" => %{
    "address" => "NotProvided",
    "postalCode" => "NotProvided",
    "securityCode" => "NotProvided",
    "status" => "NotChecked"
  },
  "bankAuthorisationCode" => "999777",
  "bankResponseCode" => "00",
  "currency" => "GBP",
  "paymentMethod" => %{
    "card" => %{
      "cardIdentifier" => "3248A1A6-5A44-470B-BC92-83536C486971",
      "cardType" => "Visa",
      "expiryDate" => "0320",
      "lastFourDigits" => "5559",
      "reusable" => false
    }
  },
  "retrievalReference" => 17717235,
  "status" => "Ok",
  "statusCode" => "0000",
  "statusDetail" => "The Authorisation was Successful.",
  "transactionId" => "EF99769B-EFE0-68EA-17B8-F38F6476B511",
  "transactionType" => "Deferred"
}
```
1. The example supplies the (billing) address to SagePay but the `avsCvcCheck` indicates otherwise. Please get this to work. Even if SagePay does not do the AVS check, you should format this into `Response.fraud_review` field.
2. The `paymentMethod` indicates that this card is NOT reusable. Surely there must be some way to generate a reusable `card_identifier`?
3. Please make use of `Response.token` field to return the card_identifier to the caller.

Reply:
curl https://pi-test.sagepay.com/api/v1/transactions \
-H "Authorization: Basic aEpZeHN3N0hMYmo0MGNCOHVkRVM4Q0RSRkxodUo4RzU0TzZyRHBVWHZFNmhZRHJyaWE6bzJpSFNyRnliWU1acG1XT1FNdWhzWFA1MlY0ZkJ0cHVTRHNocktEU1dzQlkxT2lONmh3ZDlLYjEyejRqNVVzNXU="  \
-H "Content-type: application/json" \
-X POST \
-d '{
    "transactionType": "Payment",
    "paymentMethod": {
        "card": {
            "merchantSessionKey": "<merchant-session-key>",
            "cardIdentifier": "<card-identifier>",
            "save": false
        }
    },
    "vendorTxCode": "demotransaction-<unique-suffix>",
    "amount": 10000,
    "currency": "GBP",
    "description": "Demo Payment",
    "apply3DSecure": "UseMSPSetting",
    "customerFirstName": "Sam",
    "customerLastName": "Jones",
    "billingAddress": {
        "address1": "407 St. John Street",
        "city": "London",
        "postalCode": "EC1V 4AB",
        "country": "GB"
    },
    "entryMethod": "Ecommerce"
}'




###when we used this on postman after creating merchant key and card identifier the response was


{
    "statusCode": "2007",
    "statusDetail": "Please redirect your customer to the ACSURL to complete the 3DS Transaction",
    "transactionId": "D6146D7B-033E-9E2A-C52C-87A7D3E107EA",
    "acsUrl": "https://test.sagepay.com/mpitools/accesscontroler?action=pareq",
    "paReq": "eJxVUsluwjAQvfcrIj4gXrJA0WAE5dAIaKOCVKk3y7FKVLJgJ03o19fOUoovnvc8ejPzxrBss7PzLZVOi3wxIS6eLNkDHE9Kys1BilpJBnupNf+UTposJhSTgPrBo4+9aUiIF4Q+mTCIV2/ywmAQYkbHJYBGaBSUOPG8YsDFZR29MN8PvTAANEDIpIo2jBLcn/4G1NOQ80yyg2kh5teNzIqVEEWdV4C6B+iAurIZDQGNAGp1Zk3TuLb1kl9dUbj1FyBLA7o1FNc20kamTRN2rFpSbn8+NEqa3fakd++4vkTJ61o3C0A2AxJeSWZsmGGf+g55nFM896aAOh54Zusz0vc/IChtkdXd038KjM9K5mIcYUQg27LIpcmggP5iSKQWnR2O8cPZx5EpbilAt2Genq3XojL2EWtzF1m91HhDMfY6QQsA2Vw0bBANyzbR3Sf4BaPUtJU=",
    "status": "3DAuth"
}




so this asked to go to 3D secure site and to avs-check than asked not to do 3D secure procedure so we left that part .

in transaction body we have not used "apply3DSecure": "UseMSPSetting" we have done that for without 3D secure